nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1086 @@
+                        }
+                    }
+                }?;
+
+                match region.state {
+                    RegionState::Tombstoned => Err(BackoffError::transient(
+                        WaitError::Transient(anyhow!("region not deleted yet")),
+                    )),
+
+                    RegionState::Destroyed => {
+                        info!(
+                            log,
+                            "region deleted";
+                            "region_id" => %region_id,
+                        );
+
+                        Ok(())
+                    }
+
+                    _ => Err(BackoffError::transient(WaitError::Transient(
+                        anyhow!("region unexpected state {:?}", region.state),
+                    ))),
+                }
+            },
+            |e: WaitError, delay| {
+                info!(
+                    log,
+                    "{:?}, trying again in {:?}",
+                    e,
+                    delay;
+                    "region_id" => %region_id,
+                );
+            },
+        )
+        .await
+        .map_err(|e| match e {
+            WaitError::Transient(e) => {
+                // The backoff crate can be configured with a maximum elapsed time
+                // before giving up, which means that Transient could be returned
+                // here. Our current policies do **not** set this though.
+                Error::internal_error(&e.to_string())
+            }
+
+            WaitError::Permanent(e) => e,
+        })
+    }
+
+    async fn delete_crucible_running_snapshot_impl(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+        snapshot_id: Uuid,
+    ) -> Result<(), Error> {
+        // request running snapshot deletion
+
+        self.request_crucible_running_snapshot_delete(
+            log,
+            dataset,
+            region_id,
+            snapshot_id,
+        )
+        .await?;
+
+        // `region_delete_running_snapshot` is only a request: wait until
+        // running snapshot is deleted
+        backoff::retry_notify(
+            backoff::retry_policy_internal_service_aggressive(),
+            || async {
+                let response = match self.get_crucible_region_snapshots(
+                    log,
+                    dataset,
+                    region_id,
+                )
+                .await {
+                    Ok(v) => Ok(v),
+                    Err(e) => match e {
+                        Error::Gone => {
+                            // Return Ok if the dataset's agent is gone, no
+                            // delete call is required.
+                            return Ok(());
+                        }
+
+                        _ => {
+                            Err(BackoffError::Permanent(WaitError::Permanent(e)))
+                        }
+                    }
```

Comment:
nit: this could be flattened into:
```suggestion
                    // Return Ok if the dataset's agent is gone, no
                    // delete call is required.
                    Err(Error::Gone) => return Ok(()),
                    Err(e) => Err(BackoffError::Permanent(WaitError::Permanent(e))),
```